src/algorithms/polygon_visibility/naive.rs

```diff
@@ -0,0 +1,198 @@
+//   If it leans to the right then it'll miss the 'a'-'b' edge altogether.
+//   SoS rays are essential for dealing correctly with colinear vertices.
+//
+// Note about visibility polygons:
+//   Visibility polygons are always star-shaped: All vertices are visible from the point
+//   of origin. This means that we can find the visible vertices in any order we like and
+//   then sort them around the point of origin to generate the polygon.
+//
+// Algorithm overview:
+//   1. For each vertex in the polygon (including outer boundary and any holes):
+//     1.A Shoot a right-leaning SoS ray towards the vertex and record the intersection
+//         point closest to the origin. Note: The SoS ray might not hit the vertex.
+//     1.B Shoot a left-leaning SoS ray towards the vertex and record the intersection
+//         point closest to the origin. This point may or may not be the same as the point
+//         in step 1.A.
+//   2. Sort all the intersection points counter-clockwise around the point of origin.
+//   3. If two points are at the same angle then sort according to the leaning of the
+//      ray that generated the intersection point. Ie. a point from a left-leaning ray
+//      should be Ordering::Greater than a point from a right-leaning ray.
+//
+// Pseudo code:
+//   for vertex in poly.vertex {
+//     let left_intersections = ... use left-leaning SoS ray ...;
+//     let right_intersections = ... use right-leaning SoS ray ...;
+//     let left_closest = ...;
+//     let right_closest = ...;
+//     output.push( (SoS::CounterClockWise, left_closest));
+//     output.push( (SoS::ClockWise, right_closest));
+//   }
+//   output.sort_by(|a,b| origin.ccw_cmp_around(a,b).then(cmp by sos) );
+//   Polygon::new(output)
+//
+// Test cases:
+//
+//   Input            Output
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/ /--/  |  x  \---\--/
+//  |         |     |         |
+//  \---------/     \---------/
+//
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/    |  |  x  \------\
+//  |            |  |            |
+//  \------------/  \------------/
+//
+//  /---------\     /-\     /-\
+//  |   /-\   |     |  \   /  |
+//  |   \-/   |     |   \ /   |
+//  |    x    |     |    x    |
+//  \---------/     \---------/
+
+pub fn get_visibility_polygon<T>(point: &Point<T, 2>, polygon: &Polygon<T>) -> Option<Polygon<T>>
+where
+  T: PolygonScalar,
+{
+  //check whether the point is in the polygon or not
+  if polygon.locate(point) == PointLocation::Outside {
+    return Option::None;
+  }
+
+  let polygon_segments: Vec<LineSegment<T, 2>> = polygon
+    .points
+    .windows(2)
+    .map(|pair| {
+      LineSegment::new(
+        EndPoint::Inclusive(pair[0].clone()),
+        EndPoint::Inclusive(pair[1].clone()),
+      )
+    })
+    .collect();
+
+  let mut polygon_points = Vec::new();
+  for vertex in &polygon.points {
```

Comment:
Turns out that sorting the vertices before finding intersections is much easier than sorting afterward.
Try something like this:
```rust
let mut vertices = polygon.iter_boundary().collect();
vertices.sort_by(|a,b| point.ccw_cmp_around(a,b) );
for vertex in vertices {
...
```
Note: It'll change the type of `vertex` to `Cursor<'_, T>'. You'll have to change the code in a few places to make this work.